### PR TITLE
0.4.0

### PR DESCRIPTION
**Pull request by tagdapix**
This merges in Andrey's changes to allow using captcha from geetest, and using alternate authorization.

### CODE REVIEW

**Comment by tagdapix** [comment]
Note that geetest requires some extra permissions when using content-security-policy:

Content-Security-Policy "default-src 'self'; script-src api.geetest.com monitor.geetest.com dn-staticdown.qbox.me static.geetest.com 'self' 'unsafe-eval'; style-src static.geetest.com dn-staticdown.qbox.me 'self'; connect-src api.commerce.coinbase.com  api.geetest.com 'self'; img-src data: static.geetest.com 'self'"

The unsafe-eval I'm not entirely happy about, there's really no excuse for this captcha to be using jsonp. I'll leave final decision on merge up to @lukestokes

**Comment by lukestokes** [comment]
Well, I built an entire business (FoxyCart) that dealt with credit cards using jsonp, and we didn't have too many complaints. There's no personally identifiable information involved in the free giveaway process so I'm going to call this good and move forward. If there are better options, please open an issue ticket and we can revisit.

File: src/components/FormAccountReg.vue [hunk 1]
```diff
@@ -97,31 +107,93 @@ export default {
...
         const {address, publicKey} = this
-        this.$store.dispatch('Account/isAccountRegistered', {address, publicKey})
+        const selectedDomain = address.split('@')[1]
+        const domainLimit = this.Wallet.wallet.domains_limit[selectedDomain] || {}
+        const registered = this.Wallet.wallet.accountsByDomain[selectedDomain] || 0
+        if (registered >= parseInt(domainLimit)) {
+          return this.limitError = true
+        }
+        this.$store.dispatch('Account/isAccountRegistered', {address, publicKey, cb: isRegistered => {
+          if (this.buyAddress && this.freeSale && !isRegistered) {
```
Comment (lukestokes) [inline src/components/FormAccountReg.vue:129]: Possible place we can modify to disable captcha

File: server/routes/wallet.js [hunk 2]
```diff
@@ -233,20 +259,35 @@ router.post('/public-api/buy-address', handler(async (req, res) => {
...
     return res.status(404).send({error: `Already registered`})
   }
 
   const {name, logo_url} = wallet
   const price = +Number(wallet[`${type}_sale_price`])
 
-  if(price === 0) {
-    if(!res.state.user_id) {
+  if (price === 0) {
+    let isCaptchaSuccess = false
```
Comment (lukestokes) [inline server/routes/wallet.js:279]: Possible place we can modify to disable captcha

File: src/components/FormAccountReg.vue [hunk 3]
```diff
@@ -97,31 +107,93 @@ export default {
...
+            this.initCaptcha()
+          }
+        }})
       } else {
-        const {referralCode, address, publicKey} = this
-        const redirectUrl = window.location.href
+        if (this.buyAddress && this.freeSale) {
+          if (!this.captchaObj) return
```
Comment (lukestokes) [inline src/components/FormAccountReg.vue:135]: Possible place we can modify to disable captcha

File: src/components/FormAccountReg.vue [hunk 4]
```diff
@@ -97,31 +107,93 @@ export default {
...
+        }
+        this.$store.dispatch('Account/isAccountRegistered', {address, publicKey, cb: isRegistered => {
+          if (this.buyAddress && this.freeSale && !isRegistered) {
+            this.initCaptcha()
+          }
+        }})
       } else {
-        const {referralCode, address, publicKey} = this
-        const redirectUrl = window.location.href
+        if (this.buyAddress && this.freeSale) {
```
Comment (lukestokes) [inline src/components/FormAccountReg.vue:134]: Possible place we can modify to disable captcha

File: src/components/FormAccountReg.vue [hunk 5]
```diff
@@ -97,31 +107,93 @@ export default {
...
+      const { referralCode, address, publicKey } = this
+      const redirectUrl = window.location.href
+
+      this.$store.dispatch('Server/post', {
+        key: 'buyResult', path: '/public-api/buy-address',
+        body: { address, referralCode, publicKey, redirectUrl, ...captchaParams }
+      })
+    },
+
+    initCaptcha() {
```
Comment (lukestokes) [inline src/components/FormAccountReg.vue:170]: Possible place we can modify to disable captcha (just return)

File: src/components/FormAccountReg.vue [hunk 6]
```diff
@@ -97,31 +107,93 @@ export default {
...
+        key: 'getCaptchaResult', path: '/public-api/gt/register-slide'
+      })
     }
   },
 
   watch: {
     ['buyResult._loading']: function(loading) {
-      if(loading) { return }
+      if (loading) { return }
+      if (this.buyResult.captchaStatus === 'fail') {
```
Comment (lukestokes) [inline src/components/FormAccountReg.vue:190]: Possible place we can modify to disable captcha

File: src/components/FormAccountReg.vue [hunk 7]
```diff
@@ -142,6 +214,31 @@ export default {
...
+      this.captchaLoading = true
+      if (loading) return
+      const data = this.getCaptchaResult
+      if (!data.success) {
+        this.captchaObj = null
+        this.captchaLoading = false
+        this.captchaErrored = true
+        return
+      }
+      window.initGeetest({
```
Comment (lukestokes) [inline src/components/FormAccountReg.vue:228]: Possible place we can modify to disable captcha

File: src/components/FormAccountReg.vue [hunk 8]
```diff
@@ -169,6 +266,10 @@ export default {
         this.address === this.Account.availableAccount
     },
 
+    captchaIsLoading() {
+      return this.captchaLoading
```
Comment (lukestokes) [inline src/components/FormAccountReg.vue:270]: Possible place we can modify to disable captcha

File: src/components/FormAccountReg.vue [hunk 9]
```diff
@@ -193,10 +298,14 @@ export default {
         return {error: `${type} "${this.address}" is already registered`}
       }
 
-      if(this.buyResult.error) {
+      if (this.buyResult.error) {
         return {error: this.buyResult.error}
       }
 
+      if (this.captchaErrored) {
+        return { error: 'Temporarily unavailable' }
```
Comment (lukestokes) [inline src/components/FormAccountReg.vue:306]: Possible place we can modify to disable captcha

File: src/components/FormAccountReg.vue [hunk 10]
```diff
@@ -206,6 +315,10 @@ export default {
       return {}
     },
 
+    isCaptchaInitSuccess() {
+      return !this.captchaErrored
```
Comment (lukestokes) [inline src/components/FormAccountReg.vue:319]: Possible place we can modify to disable captcha

**Pull request merged**